File: snippets/screens/compose/filter_screen/3/filter_screen.kt
package com.example.compose.search

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import java.text.SimpleDateFormat
import java.util.*

data class FilterState(
    val selectedCategories: Set<String> = emptySet(),
    val selectedTags: Set<String> = emptySet(),
    val selectedFileTypes: Set<String> = emptySet(),
    val priceRange: ClosedFloatingPointRange<Float> = 0f..1000f,
    val ratingRange: ClosedFloatingPointRange<Float> = 0f..5f,
    val selectedSortBy: String = "relevance",
    val inStock: Boolean = false,
    val freeShipping: Boolean = false,
    val onSale: Boolean = false,
    val fromDate: Long? = null,
    val toDate: Long? = null
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun FilterScreen(
    initialFilters: FilterState = FilterState(),
    onApplyFilters: (FilterState) -> Unit = {},
    onDismiss: () -> Unit = {}
) {
    var filterState by remember { mutableStateOf(initialFilters) }

    val availableCategories = listOf(
        "Electronics", "Clothing", "Books", "Home & Garden",
        "Sports", "Beauty", "Automotive", "Toys"
    )

    val availableTags = listOf(
        "Popular", "Trending", "New", "Featured",
        "Best Seller", "Limited Edition", "Premium", "Eco-Friendly"
    )

    val availableFileTypes = listOf(
        "PDF", "DOC", "XLS", "PPT", "IMG", "VID", "AUD", "ZIP"
    )

    val sortOptions = listOf(
        "relevance" to "Relevance",
        "price_low" to "Price: Low to High",
        "price_high" to "Price: High to Low",
        "rating" to "Highest Rated",
        "newest" to "Newest",
        "popularity" to "Most Popular"
    )

    fun resetFilters() {
        filterState = FilterState()
    }

    val activeFiltersCount = with(filterState) {
        var count = 0
        count += selectedCategories.size
        count += selectedTags.size
        count += selectedFileTypes.size
        if (priceRange.start > 0f || priceRange.endInclusive < 1000f) count++
        if (ratingRange.start > 0f || ratingRange.endInclusive < 5f) count++
        if (selectedSortBy != "relevance") count++
        if (inStock) count++
        if (freeShipping) count++
        if (onSale) count++
        if (fromDate != null || toDate != null) count++
        count
    }

    // Bottom Sheet Layout
    Column(
        modifier = Modifier.fillMaxSize()
    ) {
        // Handle
        Box(
            modifier = Modifier.fillMaxWidth(),
            contentAlignment = Alignment.Center
        ) {
            Surface(
                modifier = Modifier
                    .size(width = 40.dp, height = 4.dp)
                    .padding(vertical = 8.dp),
                shape = MaterialTheme.shapes.small,
                color = MaterialTheme.colorScheme.outline
            ) {}
        }

        // Header
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 16.dp, vertical = 8.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = "Filters",
                    style = MaterialTheme.typography.titleLarge,
                    fontWeight = FontWeight.Bold
                )
                if (activeFiltersCount > 0) {
                    Spacer(modifier = Modifier.width(8.dp))
                    Badge {
                        Text(activeFiltersCount.toString())
                    }
                }
            }

            TextButton(onClick = ::resetFilters) {
                Text("Reset")
            }
        }

        // Filter Content
        LazyColumn(
            modifier = Modifier.weight(1f),
            contentPadding = PaddingValues(horizontal = 16.dp),
            verticalArrangement = Arrangement.spacedBy(24.dp)
        ) {
            // Categories
            item {
                FilterSection(title = "Categories") {
                    LazyRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        items(availableCategories) { category ->
                            FilterChip(
                                onClick = {
                                    filterState = if (filterState.selectedCategories.contains(category)) {
                                        filterState.copy(selectedCategories = filterState.selectedCategories - category)
                                    } else {
                                        filterState.copy(selectedCategories = filterState.selectedCategories + category)
                                    }
                                },
                                label = { Text(category) },
                                selected = filterState.selectedCategories.contains(category)
                            )
                        }
                    }
                }
            }

            // Tags
            item {
                FilterSection(title = "Tags") {
                    LazyRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        items(availableTags) { tag ->
                            FilterChip(
                                onClick = {
                                    filterState = if (filterState.selectedTags.contains(tag)) {
                                        filterState.copy(selectedTags = filterState.selectedTags - tag)
                                    } else {
                                        filterState.copy(selectedTags = filterState.selectedTags + tag)
                                    }
                                },
                                label = { Text(tag) },
                                selected = filterState.selectedTags.contains(tag)
                            )
                        }
                    }
                }
            }

            // File Types
            item {
                FilterSection(title = "File Types") {
                    LazyRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        items(availableFileTypes) { type ->
                            FilterChip(
                                onClick = {
                                    filterState = if (filterState.selectedFileTypes.contains(type)) {
                                        filterState.copy(selectedFileTypes = filterState.selectedFileTypes - type)
                                    } else {
                                        filterState.copy(selectedFileTypes = filterState.selectedFileTypes + type)
                                    }
                                },
                                label = { Text(type) },
                                selected = filterState.selectedFileTypes.contains(type)
                            )
                        }
                    }
                }
            }

            // Price Range
            item {
                FilterSection(
                    title = "Price Range ($${filterState.priceRange.start.toInt()} - $${filterState.priceRange.endInclusive.toInt()})"
                ) {
                    RangeSlider(
                        value = filterState.priceRange,
                        onValueChange = { filterState = filterState.copy(priceRange = it) },
                        valueRange = 0f..1000f,
                        steps = 19
                    )
                }
            }

            // Rating Range
            item {
                FilterSection(
                    title = "Minimum Rating (${String.format("%.1f", filterState.ratingRange.start)} stars)"
                ) {
                    Slider(
                        value = filterState.ratingRange.start,
                        onValueChange = {
                            filterState = filterState.copy(ratingRange = it..filterState.ratingRange.endInclusive)
                        },
                        valueRange = 0f..5f,
                        steps = 9
                    )
                }
            }

            // Sort By
            item {
                FilterSection(title = "Sort By") {
                    var expanded by remember { mutableStateOf(false) }
                    ExposedDropdownMenuBox(
                        expanded = expanded,
                        onExpandedChange = { expanded = it }
                    ) {
                        OutlinedTextField(
                            value = sortOptions.find { it.first == filterState.selectedSortBy }?.second ?: "Relevance",
                            onValueChange = { },
                            readOnly = true,
                            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
                            modifier = Modifier
                                .fillMaxWidth()
                                .menuAnchor()
                        )
                        ExposedDropdownMenu(
                            expanded = expanded,
                            onDismissRequest = { expanded = false }
                        ) {
                            sortOptions.forEach { (value, label) ->
                                DropdownMenuItem(
                                    text = { Text(label) },
                                    onClick = {
                                        filterState = filterState.copy(selectedSortBy = value)
                                        expanded = false
                                    }
                                )
                            }
                        }
                    }
                }
            }

            // Boolean Options
            item {
                FilterSection(title = "Options") {
                    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Switch(
                                checked = filterState.inStock,
                                onCheckedChange = { filterState = filterState.copy(inStock = it) }
                            )
                            Spacer(modifier = Modifier.width(16.dp))
                            Text("In Stock Only")
                        }

                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Switch(
                                checked = filterState.freeShipping,
                                onCheckedChange = { filterState = filterState.copy(freeShipping = it) }
                            )
                            Spacer(modifier = Modifier.width(16.dp))
                            Text("Free Shipping")
                        }

                        Row(
                            modifier = Modifier.fillMaxWidth(),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Switch(
                                checked = filterState.onSale,
                                onCheckedChange = { filterState = filterState.copy(onSale = it) }
                            )
                            Spacer(modifier = Modifier.width(16.dp))
                            Text("On Sale")
                        }
                    }
                }
            }

            // Date Range
            item {
                FilterSection(title = "Date Range") {
                    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                        OutlinedTextField(
                            value = filterState.fromDate?.let {
                                SimpleDateFormat("MMM dd, yyyy", Locale.getDefault()).format(Date(it))
                            } ?: "",
                            onValueChange = { },
                            label = { Text("From Date") },
                            readOnly = true,
                            trailingIcon = {
                                IconButton(onClick = { /* Open date picker */ }) {
                                    Icon(Icons.Default.DateRange, contentDescription = "Select date")
                                }
                            },
                            modifier = Modifier.fillMaxWidth()
                        )

                        OutlinedTextField(
                            value = filterState.toDate?.let {
                                SimpleDateFormat("MMM dd, yyyy", Locale.getDefault()).format(Date(it))
                            } ?: "",
                            onValueChange = { },
                            label = { Text("To Date") },
                            readOnly = true,
                            trailingIcon = {
                                IconButton(onClick = { /* Open date picker */ }) {
                                    Icon(Icons.Default.DateRange, contentDescription = "Select date")
                                }
                            },
                            modifier = Modifier.fillMaxWidth()
                        )
                    }
                }
            }

            // Bottom spacing for button
            item {
                Spacer(modifier = Modifier.height(80.dp))
            }
        }

        // Apply Button
        Surface(
            shadowElevation = 8.dp
        ) {
            Button(
                onClick = { onApplyFilters(filterState) },
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp)
            ) {
                Text(
                    if (activeFiltersCount > 0)
                        "Apply $activeFiltersCount Filters"
                    else
                        "Apply Filters"
                )
            }
        }
    }
}

@Composable
private fun FilterSection(
    title: String,
    content: @Composable ColumnScope.() -> Unit
) {
    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(
            text = title,
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.Bold
        )
        content()
    }
}

@Preview(showBackground = true)
@Composable
private fun FilterScreenPreview() {
    MaterialTheme {
        FilterScreen()
    }
}

File: snippets/screens/flutter/filter_screen/3/filter_screen.dart
import 'package:flutter/material.dart';

class FilterScreen extends StatefulWidget {
  final Map<String, dynamic>? initialFilters;

  const FilterScreen({super.key, this.initialFilters});

  @override
  State<FilterScreen> createState() => _FilterScreenState();
}

class _FilterScreenState extends State<FilterScreen> {
  final Set<String> _selectedCategories = {};
  final Set<String> _selectedTags = {};
  final Set<String> _selectedFileTypes = {};
  RangeValues _priceRange = const RangeValues(0, 1000);
  RangeValues _ratingRange = const RangeValues(0, 5);
  String _sortBy = 'relevance';
  bool _inStock = false;
  bool _freeShipping = false;
  bool _onSale = false;
  DateTime? _fromDate;
  DateTime? _toDate;

  final List<String> _availableCategories = [
    'Electronics',
    'Clothing',
    'Books',
    'Home & Garden',
    'Sports',
    'Beauty',
    'Automotive',
    'Toys',
  ];

  final List<String> _availableTags = [
    'Popular',
    'Trending',
    'New',
    'Featured',
    'Best Seller',
    'Limited Edition',
    'Premium',
    'Eco-Friendly',
  ];

  final List<String> _availableFileTypes = [
    'PDF',
    'DOC',
    'XLS',
    'PPT',
    'IMG',
    'VID',
    'AUD',
    'ZIP',
  ];

  @override
  void initState() {
    super.initState();
    _loadInitialFilters();
  }

  void _loadInitialFilters() {
    if (widget.initialFilters != null) {
      final filters = widget.initialFilters!;
      setState(() {
        _selectedCategories.addAll(
          List<String>.from(filters['categories'] ?? []),
        );
        _selectedTags.addAll(
          List<String>.from(filters['tags'] ?? []),
        );
        _selectedFileTypes.addAll(
          List<String>.from(filters['fileTypes'] ?? []),
        );
        _priceRange = RangeValues(
          filters['priceMin']?.toDouble() ?? 0,
          filters['priceMax']?.toDouble() ?? 1000,
        );
        _ratingRange = RangeValues(
          filters['ratingMin']?.toDouble() ?? 0,
          filters['ratingMax']?.toDouble() ?? 5,
        );
        _sortBy = filters['sortBy'] ?? 'relevance';
        _inStock = filters['inStock'] ?? false;
        _freeShipping = filters['freeShipping'] ?? false;
        _onSale = filters['onSale'] ?? false;
        _fromDate = filters['fromDate'];
        _toDate = filters['toDate'];
      });
    }
  }

  void _applyFilters() {
    final filters = {
      'categories': _selectedCategories.toList(),
      'tags': _selectedTags.toList(),
      'fileTypes': _selectedFileTypes.toList(),
      'priceMin': _priceRange.start,
      'priceMax': _priceRange.end,
      'ratingMin': _ratingRange.start,
      'ratingMax': _ratingRange.end,
      'sortBy': _sortBy,
      'inStock': _inStock,
      'freeShipping': _freeShipping,
      'onSale': _onSale,
      'fromDate': _fromDate,
      'toDate': _toDate,
    };

    Navigator.pop(context, filters);
  }

  void _resetFilters() {
    setState(() {
      _selectedCategories.clear();
      _selectedTags.clear();
      _selectedFileTypes.clear();
      _priceRange = const RangeValues(0, 1000);
      _ratingRange = const RangeValues(0, 5);
      _sortBy = 'relevance';
      _inStock = false;
      _freeShipping = false;
      _onSale = false;
      _fromDate = null;
      _toDate = null;
    });
  }

  int get _activeFiltersCount {
    int count = 0;
    count += _selectedCategories.length;
    count += _selectedTags.length;
    count += _selectedFileTypes.length;
    if (_priceRange.start > 0 || _priceRange.end < 1000) count++;
    if (_ratingRange.start > 0 || _ratingRange.end < 5) count++;
    if (_sortBy != 'relevance') count++;
    if (_inStock) count++;
    if (_freeShipping) count++;
    if (_onSale) count++;
    if (_fromDate != null || _toDate != null) count++;
    return count;
  }

  @override
  Widget build(BuildContext context) {
    return DraggableScrollableSheet(
      initialChildSize: 0.9,
      minChildSize: 0.5,
      maxChildSize: 0.95,
      builder: (context, scrollController) {
        return Container(
          decoration: const BoxDecoration(
            color: Colors.white,
            borderRadius: BorderRadius.vertical(top: Radius.circular(20)),
          ),
          child: Column(
            children: [
              // Handle
              Container(
                margin: const EdgeInsets.symmetric(vertical: 8),
                height: 4,
                width: 40,
                decoration: BoxDecoration(
                  color: Colors.grey[300],
                  borderRadius: BorderRadius.circular(2),
                ),
              ),

              // Header
              Padding(
                padding: const EdgeInsets.all(16),
                child: Row(
                  children: [
                    const Text(
                      'Filters',
                      style: TextStyle(
                        fontSize: 20,
                        fontWeight: FontWeight.bold,
                      ),
                    ),
                    if (_activeFiltersCount > 0) ...[
                      const SizedBox(width: 8),
                      Container(
                        padding: const EdgeInsets.symmetric(
                          horizontal: 8,
                          vertical: 4,
                        ),
                        decoration: BoxDecoration(
                          color: Theme.of(context).primaryColor,
                          borderRadius: BorderRadius.circular(12),
                        ),
                        child: Text(
                          _activeFiltersCount.toString(),
                          style: const TextStyle(
                            color: Colors.white,
                            fontSize: 12,
                            fontWeight: FontWeight.bold,
                          ),
                        ),
                      ),
                    ],
                    const Spacer(),
                    TextButton(
                      onPressed: _resetFilters,
                      child: const Text('Reset'),
                    ),
                  ],
                ),
              ),

              // Filter Content
              Expanded(
                child: ListView(
                  controller: scrollController,
                  padding: const EdgeInsets.symmetric(horizontal: 16),
                  children: [
                    // Categories
                    _buildFilterSection(
                      'Categories',
                      Wrap(
                        spacing: 8,
                        children: _availableCategories
                            .map((category) => FilterChip(
                                  label: Text(category),
                                  selected:
                                      _selectedCategories.contains(category),
                                  onSelected: (selected) {
                                    setState(() {
                                      if (selected) {
                                        _selectedCategories.add(category);
                                      } else {
                                        _selectedCategories.remove(category);
                                      }
                                    });
                                  },
                                ))
                            .toList(),
                      ),
                    ),

                    // Tags
                    _buildFilterSection(
                      'Tags',
                      Wrap(
                        spacing: 8,
                        children: _availableTags
                            .map((tag) => FilterChip(
                                  label: Text(tag),
                                  selected: _selectedTags.contains(tag),
                                  onSelected: (selected) {
                                    setState(() {
                                      if (selected) {
                                        _selectedTags.add(tag);
                                      } else {
                                        _selectedTags.remove(tag);
                                      }
                                    });
                                  },
                                ))
                            .toList(),
                      ),
                    ),

                    // File Types
                    _buildFilterSection(
                      'File Types',
                      Wrap(
                        spacing: 8,
                        children: _availableFileTypes
                            .map((type) => FilterChip(
                                  label: Text(type),
                                  selected: _selectedFileTypes.contains(type),
                                  onSelected: (selected) {
                                    setState(() {
                                      if (selected) {
                                        _selectedFileTypes.add(type);
                                      } else {
                                        _selectedFileTypes.remove(type);
                                      }
                                    });
                                  },
                                ))
                            .toList(),
                      ),
                    ),

                    // Price Range
                    _buildFilterSection(
                      'Price Range (\$${_priceRange.start.round()} - \$${_priceRange.end.round()})',
                      RangeSlider(
                        values: _priceRange,
                        min: 0,
                        max: 1000,
                        divisions: 20,
                        onChanged: (values) {
                          setState(() => _priceRange = values);
                        },
                      ),
                    ),

                    // Rating Range
                    _buildFilterSection(
                      'Minimum Rating (${_ratingRange.start.toStringAsFixed(1)} stars)',
                      Slider(
                        value: _ratingRange.start,
                        min: 0,
                        max: 5,
                        divisions: 10,
                        onChanged: (value) {
                          setState(() {
                            _ratingRange = RangeValues(value, _ratingRange.end);
                          });
                        },
                      ),
                    ),

                    // Sort By
                    _buildFilterSection(
                      'Sort By',
                      DropdownButtonFormField<String>(
                        value: _sortBy,
                        decoration: const InputDecoration(
                          border: OutlineInputBorder(),
                        ),
                        items: const [
                          DropdownMenuItem(
                            value: 'relevance',
                            child: Text('Relevance'),
                          ),
                          DropdownMenuItem(
                            value: 'price_low',
                            child: Text('Price: Low to High'),
                          ),
                          DropdownMenuItem(
                            value: 'price_high',
                            child: Text('Price: High to Low'),
                          ),
                          DropdownMenuItem(
                            value: 'rating',
                            child: Text('Highest Rated'),
                          ),
                          DropdownMenuItem(
                            value: 'newest',
                            child: Text('Newest'),
                          ),
                          DropdownMenuItem(
                            value: 'popularity',
                            child: Text('Most Popular'),
                          ),
                        ],
                        onChanged: (value) {
                          setState(() => _sortBy = value!);
                        },
                      ),
                    ),

                    // Boolean Filters
                    _buildFilterSection(
                      'Options',
                      Column(
                        children: [
                          SwitchListTile(
                            title: const Text('In Stock Only'),
                            value: _inStock,
                            onChanged: (value) {
                              setState(() => _inStock = value);
                            },
                          ),
                          SwitchListTile(
                            title: const Text('Free Shipping'),
                            value: _freeShipping,
                            onChanged: (value) {
                              setState(() => _freeShipping = value);
                            },
                          ),
                          SwitchListTile(
                            title: const Text('On Sale'),
                            value: _onSale,
                            onChanged: (value) {
                              setState(() => _onSale = value);
                            },
                          ),
                        ],
                      ),
                    ),

                    // Date Range
                    _buildFilterSection(
                      'Date Range',
                      Column(
                        children: [
                          ListTile(
                            title: Text(
                              _fromDate == null
                                  ? 'From Date'
                                  : 'From: ${_fromDate!.toString().split(' ')[0]}',
                            ),
                            trailing: const Icon(Icons.date_range),
                            onTap: () => _selectDate(true),
                          ),
                          ListTile(
                            title: Text(
                              _toDate == null
                                  ? 'To Date'
                                  : 'To: ${_toDate!.toString().split(' ')[0]}',
                            ),
                            trailing: const Icon(Icons.date_range),
                            onTap: () => _selectDate(false),
                          ),
                        ],
                      ),
                    ),

                    const SizedBox(height: 100), // Space for apply button
                  ],
                ),
              ),

              // Apply Button
              Container(
                padding: const EdgeInsets.all(16),
                decoration: BoxDecoration(
                  color: Colors.white,
                  boxShadow: [
                    BoxShadow(
                      color: Colors.black.withOpacity(0.1),
                      blurRadius: 4,
                      offset: const Offset(0, -2),
                    ),
                  ],
                ),
                child: SafeArea(
                  child: SizedBox(
                    width: double.infinity,
                    child: ElevatedButton(
                      onPressed: _applyFilters,
                      style: ElevatedButton.styleFrom(
                        padding: const EdgeInsets.all(16),
                        shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(8),
                        ),
                      ),
                      child: Text(
                        _activeFiltersCount > 0
                            ? 'Apply $_activeFiltersCount Filters'
                            : 'Apply Filters',
                        style: const TextStyle(fontSize: 16),
                      ),
                    ),
                  ),
                ),
              ),
            ],
          ),
        );
      },
    );
  }

  Widget _buildFilterSection(String title, Widget content) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const SizedBox(height: 24),
        Text(
          title,
          style: const TextStyle(
            fontSize: 16,
            fontWeight: FontWeight.bold,
          ),
        ),
        const SizedBox(height: 12),
        content,
      ],
    );
  }

  Future<void> _selectDate(bool isFromDate) async {
    final picked = await showDatePicker(
      context: context,
      initialDate: DateTime.now(),
      firstDate: DateTime.now().subtract(const Duration(days: 365 * 10)),
      lastDate: DateTime.now(),
    );

    if (picked != null) {
      setState(() {
        if (isFromDate) {
          _fromDate = picked;
        } else {
          _toDate = picked;
        }
      });
    }
  }
}